src/metatensor/models/utils/device.py

```diff
@@ -0,0 +1,115 @@
+            "cpu, cuda, mps, gpu, multi-gpu, multi-cuda"
+        )
+
+    # we convert "gpu" and "multi-gpu" to "cuda" or "mps" if available
+    if requested_device == "gpu":
+        if torch.cuda.is_available():
+            requested_device = "cuda"
+        elif torch.backends.mps.is_built() and torch.backends.mps.is_available():
+            requested_device = "mps"
+        else:
+            raise ValueError(
+                "Requested `gpu` device, but found no GPU (CUDA or MPS) devices"
+            )
+
+    # we convert "multi-gpu" to "multi-cuda" if available
+    if requested_device == "multi-gpu":
+        if torch.cuda.is_available():
+            requested_device = "multi-cuda"
+        else:
+            raise ValueError("Requested `multi-gpu` device, but found no CUDA devices")
+
+    # check that the requested device is available
+    available_device_types = [device.type for device in available_devices]
+    available_device_strings = ["cpu"]  # always available
+    if "cuda" in available_device_types:
+        available_device_strings.append("cuda")
+    if "mps" in available_device_types:
+        available_device_strings.append("mps")
+    if available_device_strings.count("cuda") > 1:
+        available_device_strings.append("multi-cuda")
+
+    if requested_device not in available_device_strings:
+        if requested_device == "multi-cuda":
+            raise ValueError(
+                "Requested device `multi-gpu` or `multi-cuda`, but found only "
+                f"{available_device_strings.count('cuda')} cuda devices"
+            )
+        else:
+            raise ValueError(
+                f"Requested device {requested_device} is not available on this system"
+            )
+
+    # if the requested device is available, check it against the architecture's devices
+    if requested_device not in architecture_devices:
+        raise ValueError(
+            f"The requested device `{requested_device}` is not supported by the chosen "
+            f"architecture. Supported devices are {architecture_devices}."
+        )
+
+    # we check all the devices that come before the requested one in the
+    # list of architecture devices. If any of them are available, we warn
+
+    requested_device_index = architecture_devices.index(requested_device)
+    for device in architecture_devices[:requested_device_index]:
+        if device in available_device_strings:
+            warnings.warn(
+                f"Device `{requested_device}` was requested, but the chosen"
+                f"architecture prefers `{device}`, which was also found on your "
+                f"system. Consider using the `{device}` device.",
+                stacklevel=2,
+            )
+
+    # finally, we convert the requested device to a list of devices
+    if requested_device == "multi-cuda":
+        return [torch.device(f"cuda:{i}") for i in range(torch.cuda.device_count())]
```

Comment:
does this properly handle cases where there a 4 GPU available, but the user only selected 2 of them with an environment variable like `CUDA_VISIBLE_DEVICES`?

Reply:
You're right. I'll go for something more precise

Reply:
Actually I couldn't... all I can find online is to do what I did. I'm not sure if it works if some devices are hidden the way you said